src/dataServices/dynamoDbBundleService.ts

```diff
@@ -270,6 +325,98 @@ export class DynamoDbBundleService implements Bundle {
         }
     }
 
+    private async updatedReferences(
+        requests: BatchReadWriteRequest[],
+        lockedItems: ItemRequest[],
+    ): Promise<{ status: string; numErrors: number }> {
+        const result: { status: string; numErrors: number } = { status: 'OK', numErrors: 0 };
+        const idToVersionId: Record<string, string> = {};
+        lockedItems.forEach((itemRequest: ItemRequest) => {
+            if (itemRequest.operation === 'update' && itemRequest.vid) {
+                idToVersionId[`${itemRequest.resourceType}_${itemRequest.id}`] = `${itemRequest.vid + 1}`;
+            }
+        });
+
+        const createOrUpdates = requests.filter((request: BatchReadWriteRequest) => {
+            return request.operation === 'create' || request.operation === 'update';
+        });
+
+        requests.forEach((request: BatchReadWriteRequest) => {
+            const key = `${request.resourceType}_${request.id}`;
+            if (request.operation === 'create') {
+                idToVersionId[key] = '1';
+            }
+            // Setting version id to '1' of resources in the bundle that have not been locked. because
+            // if updateCreateSupported==true creates may come disguised as updates. During locking they obviously weren't found
+            // now we don't want to search for them again and then fail because we can't find them.
+            if (request.operation === 'update' && this.updateCreateSupported && !(key in idToVersionId)) {
+                idToVersionId[key] = '1';
+            }
+        });
+
+        await Promise.all(
+            createOrUpdates.map(async (request: BatchReadWriteRequest) => {
```

Comment:
I think that this section can be simplified.

The code is hard to follow since there are nested `Promise.all` that are mostly just executing synchronous code. The only truly async part is the call to DDB.

I suggest to refactor to do the synchronous stuff first with regular for loops or array functions, and only use `Promise.all` for the call to DDB

Maybe something like:
```js
const requestsWithReferencesThatMustBeLookedUp = createOrUpdates
    .filter(resourceTypeInVersionedLinksArray())
    .flatMap(request => (extractReferences()))
    .filter(referenceInVersionedLinksArray());

// requestsWithReferencesThatMustBeLookedUp is something like
// [
//     {
//         request: '<original request object>',
//         reference: {
//             path: 'a.b.c',
//             value: 'val'
//         },
//     },
//     ...
// ]

const requestsForDDB = [];

for (const requestObj of requestsWithReferencesThatMustBeLookedUp) {
    vid = findInIdToVersionId(requestObj)
    
    if (vid){
        // set reference to `/_history/${vid}`
    } else {
        requestsForDDB.push(requestObj);
    }
}

const ddbResults = await Promise.all(requestsForDDB)

for (const ddbResult of ddbResults) {
    // set reference to `/_history/${vid}`
}
```

Reply:
These are great suggestions. Thank you.